iodata/formats/gromacs.py

```diff
@@ -0,0 +1,145 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, see <http://www.gnu.org/licenses/>
+# --
+"""gro file format.
+
+Files with the gro file extension contain a molecular structure in Gromos87 format.
+gro files can be used as trajectory by simply concatenating files.
+
+http://manual.gromacs.org/current/reference-manual/file-formats.html#gro
+
+"""
+
+
+from typing import Tuple, Iterator
+
+import numpy as np
+
+from ..docstrings import (document_load_one, document_load_many)
+from ..utils import nanometer, picosecond, LineIterator
+
+
+__all__ = []
+
+
+PATTERNS = ['*.gro']
+
+
+@document_load_one('GRO', ['atcoords', 'atffparams', 'cellvecs', 'extra', 'title'])
+def load_one(lit: LineIterator) -> dict:
+    """Do not edit this docstring. It will be overwritten."""
+    while True:
+        try:
+            data = _helper_read_frame(lit)
+        except StopIteration:
+            break
+        if data is not None:
+            title = data[0]
+            time = data[1]
+            resnums = np.array(data[2])
+            resnames = np.array(data[3])
+            attypes = np.array(data[4])
+            atcoords = data[5]
+            velocities = data[6]
+            cellvecs = data[7]
+            atffparams = {
+                'attypes': attypes,
+                'resnames': resnames,
+                'resnums': resnums
+            }
+            extra = {
+                'time': time,
+                'velocities': velocities
+            }
+            result = {
+                'atcoords': atcoords,
+                'atffparams': atffparams,
+                'cellvecs': cellvecs,
+                'extra': extra,
+                'title': title,
+            }
+            return result
+    raise lit.error('Bad format. Gromacs gro file could not be read.')
+
+
+@document_load_many('GRO', ['atcoords', 'atffparams', 'cellvecs', 'extra', 'title'])
+def load_many(lit: LineIterator) -> Iterator[dict]:
+    """Do not edit this docstring. It will be overwritten."""
+    # gro files can be used as trajectory by simply concatenating files,
+    # making it trivial to load many frames.
+    while True:
+        try:
+            yield load_one(lit)
+        except IOError:
+            return
+
+
+def _helper_read_frame(lit: LineIterator) -> Tuple:
+    """Read one frame."""
+    # Read the first line, get the title and try to get the time.
+    # Time field is optional.
+    line = next(lit)
+    title = line.split(',')[0] if 't=' in line else line[:-1]
+    pos = line.rfind('t=')
+    if pos >= 0:
+        time = float(line[pos + 2:]) * picosecond
+    else:
+        time = 0.0
+    # Read the second line for number of atoms.
+    natoms = next(lit)
+    if natoms is not None:
+        try:
+            natoms = int(natoms)
+        except ValueError:
+            return None
```

Comment:
Can you add a comment in the code to explain why you need a `try/except` statement here? 
I don't have any experience with `GRO` files, but I am thinking that `if natoms is not None`, then you can convert it into `int`... And I am not sure why there is a `return None`. Does that specify that there is a `Bad file`?

Reply:
I will add it. Regarding the second question, that is correct. This is a simple test to evaluate that the number of atoms in the file is present in the second row. If this line is anything else, we can assume that the format is incorrect. Obviously, I can add more checkpoints, for me there is no problem at all.

Reply:
These checks are good, but actually don't do checks like this because then there are so many checks that you need to include. For example, the first line of XYZ file is also supposed to be an integer (for number of atoms), but in loading an XYZ file we assume it is the case. See: https://github.com/theochem/iodata/blob/master/iodata/formats/xyz.py#L101
If an XYZ file is incorrect, and the first line isn't an integer, the `natom = int(next(lit))` would fail, and the error should be clear-ish on what went wrong.

I think it is safe to use `natom = int(next(lit))` here as well, then you can also simplify the `load_one` a bit, by removing try/except and if statements because data would not be `None`, see: https://github.com/theochem/iodata/pull/172/files#diff-e32c760d08263acbc390686b11208558R47-R51

Reply:
Thanks for the meaningful and explanatory reply! I really appreciate it, because it allows me to understand a little more about the module's work philosophy. I will commit these changes as soon as possible.

Reply:
> I think it is safe to use `natom = int(next(lit))` here as well, then you can also simplify the `load_one` a bit, by removing try/except and if statements because data would not be `None`, see: https://github.com/theochem/iodata/pull/172/files#diff-e32c760d08263acbc390686b11208558R47-R51

Finally, I removed the if statement only. I was using the try/except statement for the load_many function in case that the gro file was a trajectory.